File: protocol/test_protocol.py
"""test_protocol.py — Python-side self-test, mirrors test_protocol.c."""
import random
from protocol import Receiver, build_frame, crc16, Opcode, MsgType

fails = 0


def check(cond: bool, msg: str) -> None:
    global fails
    print(f"  [{'OK' if cond else 'FAIL'}] {msg}")
    if not cond:
        fails += 1


def test_crc_known_vector():
    print("Test 1: CRC-16/CCITT-FALSE known test vector")
    crc = crc16(b"123456789")
    print(f"  computed=0x{crc:04X} expected=0x29B1")
    check(crc == 0x29B1, "CRC matches published CRC-16/CCITT-FALSE check value")


def test_roundtrip_every_opcode():
    print("Test 2: build + decode round-trip for every opcode")
    ok = True
    for i, op in enumerate(Opcode):
        payload = bytes([op.value, 0xAB])
        frame_bytes = build_frame(MsgType.CMD, i & 0xFF, payload)
        rx = Receiver()
        frames = rx.feed(frame_bytes)
        if not (
            len(frames) == 1
            and frames[0].msg_type == MsgType.CMD
            and frames[0].msg_id == (i & 0xFF)
            and frames[0].payload == payload
        ):
            ok = False
    check(ok, "all 11 command opcodes survive build->wire->decode unchanged")


def test_corruption_rejected_and_resyncs():
    print("Test 3: corrupted frame dropped, next valid frame still decodes")
    good1 = bytearray(build_frame(MsgType.CMD, 1, bytes([0x2A])))
    good2 = build_frame(MsgType.CMD, 2, bytes([0x2A]))
    good1[1] ^= 0xFF

    rx = Receiver()
    frames = rx.feed(bytes(good1))
    check(len(frames) == 0 and rx.last_error == "crc_mismatch", "corrupted frame reported, not accepted")

    frames = rx.feed(good2)
    check(len(frames) == 1 and frames[0].msg_id == 2, "decoder resyncs on the very next frame")


def test_fuzz_never_false_accepts():
    print("Test 4: fuzz — random corruption never false-accepts")
    random.seed(42)
    payload = bytes([0x11, 0x22, 0x33, 0x44])
    false_accepts = 0
    iterations = 50000
    for i in range(iterations):
        frame = bytearray(build_frame(MsgType.CMD, i & 0xFF, payload))
        for _ in range(random.randint(0, 2)):
            idx = random.randrange(0, max(1, len(frame) - 1))
            frame[idx] ^= random.randint(1, 255)
        rx = Receiver()
        frames = rx.feed(bytes(frame))
        for f in frames:
            if not (f.msg_type == MsgType.CMD and f.msg_id == (i & 0xFF) and f.payload == payload):
                false_accepts += 1
    print(f"  {iterations} iterations, {false_accepts} false-accepts")
    check(false_accepts == 0, "no corrupted frame was ever accepted as valid")


if __name__ == "__main__":
    print("=== MedMetrix protocol.py self-test ===\n")
    test_crc_known_vector()
    test_roundtrip_every_opcode()
    test_corruption_rejected_and_resyncs()
    test_fuzz_never_false_accepts()
    status = "ALL TESTS PASSED" if fails == 0 else "TESTS FAILED"
    print(f"\n=== {status} ({fails} failing check{'s' if fails != 1 else ''}) ===")
    raise SystemExit(0 if fails == 0 else 1)

File: protocol/test_protocol.c
/* test_protocol.c — host-side test harness for mm_protocol.h. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "mm_protocol.h"

static int fails = 0;
#define CHECK(cond, msg) do { \
    if (!(cond)) { printf("  [FAIL] %s\n", msg); fails++; } \
    else { printf("  [ OK ] %s\n", msg); } \
} while (0)

static void test_crc_known_vector(void) {
    printf("Test 1: CRC-16/CCITT-FALSE known test vector\n");
    const uint8_t v[] = "123456789";
    uint16_t crc = mm_crc16(v, 9);
    printf("  computed=0x%04X expected=0x29B1\n", crc);
    CHECK(crc == 0x29B1, "CRC matches published CRC-16/CCITT-FALSE check value");
}

static void test_roundtrip_every_opcode(void) {
    printf("Test 2: build + decode round-trip for every opcode\n");
    mm_opcode_t opcodes[] = {
        MM_CMD_SET_PUMP1_RATE, MM_CMD_SET_PUMP2_RATE, MM_CMD_SET_O2_VALVE_POS,
        MM_CMD_HOME_O2_VALVE, MM_CMD_QUERY_ESTOP, MM_CMD_GET_ECG, MM_CMD_GET_IMU,
        MM_CMD_GET_SPO2, MM_CMD_GET_TEMP, MM_CMD_GET_NIBP, MM_CMD_GET_FULL_STATUS
    };
    int n = (int)(sizeof(opcodes) / sizeof(opcodes[0]));
    int all_ok = 1;

    for (int i = 0; i < n; i++) {
        uint8_t payload[2] = { (uint8_t)(opcodes[i]), 0xAB };
        uint8_t encoded[MM_MAX_ENCODED];
        size_t enc_len = mm_build_frame(MM_MSG_CMD, (uint8_t)i, payload, 2, encoded);

        mm_receiver_t rx;
        mm_receiver_init(&rx);
        mm_frame_t got;
        int result = 0;
        for (size_t b = 0; b < enc_len; b++) {
            result = mm_receiver_feed(&rx, encoded[b], &got);
        }
        int ok = (result == 1) && got.msg_type == MM_MSG_CMD && got.msg_id == (uint8_t)i
                  && got.payload_len == 2 && got.payload[0] == (uint8_t)opcodes[i] && got.payload[1] == 0xAB;
        if (!ok) all_ok = 0;
    }
    CHECK(all_ok, "all 11 command opcodes survive build->wire->decode unchanged");
}

static void test_corruption_rejected_and_resyncs(void) {
    printf("Test 3: corrupted frame is dropped, next valid frame still decodes\n");
    uint8_t payload[1] = { 0x2A };
    uint8_t good1[MM_MAX_ENCODED], good2[MM_MAX_ENCODED];
    size_t len1 = mm_build_frame(MM_MSG_CMD, 1, payload, 1, good1);
    size_t len2 = mm_build_frame(MM_MSG_CMD, 2, payload, 1, good2);

    uint8_t corrupt1[MM_MAX_ENCODED];
    memcpy(corrupt1, good1, len1);
    corrupt1[1] ^= 0xFF;

    mm_receiver_t rx;
    mm_receiver_init(&rx);
    mm_frame_t got;
    int corrupt_result = 0, resync_result = 0;

    for (size_t b = 0; b < len1; b++) corrupt_result = mm_receiver_feed(&rx, corrupt1[b], &got);
    for (size_t b = 0; b < len2; b++) resync_result = mm_receiver_feed(&rx, good2[b], &got);

    CHECK(corrupt_result == -1, "corrupted frame reported as CRC failure, not silently accepted");
    CHECK(resync_result == 1 && got.msg_id == 2, "decoder resyncs on the very next frame, no restart needed");
}

static void test_fuzz_never_crashes_or_false_accepts(void) {
    printf("Test 4: fuzz — random byte corruption never crashes and never false-accepts\n");
    srand(42);
    uint8_t payload[4] = { 0x11, 0x22, 0x33, 0x44 };
    int iterations = 200000;
    int false_accepts = 0;

    for (int iter = 0; iter < iterations; iter++) {
        uint8_t frame[MM_MAX_ENCODED];
        size_t len = mm_build_frame(MM_MSG_CMD, (uint8_t)(iter & 0xFF), payload, 4, frame);

        int n_flips = rand() % 3;
        for (int f = 0; f < n_flips; f++) {
            size_t idx = (size_t)(rand() % (int)(len > 1 ? len - 1 : 1));
            frame[idx] ^= (uint8_t)(1 + (rand() % 255));
        }

        mm_receiver_t rx;
        mm_receiver_init(&rx);
        mm_frame_t got;
        int result = 0;
        for (size_t b = 0; b < len; b++) {
            result = mm_receiver_feed(&rx, frame[b], &got);
        }

        if (result == 1) {
            int matches = got.msg_type == MM_MSG_CMD && got.msg_id == (uint8_t)(iter & 0xFF)
                          && got.payload_len == 4 && memcmp(got.payload, payload, 4) == 0;
            if (!matches) false_accepts++;
        }
    }
    printf("  %d iterations, %d false-accepts\n", iterations, false_accepts);
    CHECK(false_accepts == 0, "no corrupted frame was ever accepted as if it were valid");
}

int main(void) {
    printf("=== MedMetrix protocol library self-test ===\n\n");
    test_crc_known_vector();
    test_roundtrip_every_opcode();
    test_corruption_rejected_and_resyncs();
    test_fuzz_never_crashes_or_false_accepts();

    printf("\n=== %s (%d failing check%s) ===\n",
           fails == 0 ? "ALL TESTS PASSED" : "TESTS FAILED",
           fails, fails == 1 ? "" : "s");
    return fails == 0 ? 0 : 1;
}
